### PR TITLE
feat(tokenizer): Add Mistral and Llama support with backwards compatibility

### PR DESCRIPTION
**Pull request by rtalaricw**
This PR adds support for Mistral while maintaining backwards compatibility with Pythia models. It is sourced from the work of @Rexwang8 from [here](https://github.com/wbrown/gpt_bpe/tree/rwang.mistralsupport04132024).

### CODE REVIEW

File: gpt_bpe_test.go [hunk 1]
```diff
@@ -787,25 +790,102 @@ func TestLlamaEncoder_Encode(t *testing.T) {
...
+	var rsrcType resources.ResourceType
+	rsrcType = resources.RESOURCETYPE_TRANSFORMERS
+	hfApiToken := os.Getenv("HF_API_TOKEN")
+	os.MkdirAll(destPath, 0755)
+	_, rsrcErr := resources.ResolveResources(modelId, destPathPTR,
+		resources.RESOURCE_MODEL, rsrcType, hfApiToken)
+	if rsrcErr != nil {
+		os.RemoveAll(destPath)
+		t.Errorf("Error downloading model resources: %s", rsrcErr)
+	}
```
Comment (sangstar) [inline gpt_bpe_test.go:863]: Are you sure that extra `os.RemoveAll(destPath)` is necessary when you're deferring it in the outer scope? Is there any risk of `t.Errorf` not returning or something?

```suggestion
	if rsrcErr != nil {
		t.Errorf("Error downloading model resources: %s", rsrcErr)
	}
```

File: gpt_bpe_test.go [hunk 2]
```diff
@@ -787,25 +790,102 @@ func TestLlamaEncoder_Encode(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("Error creating encoder: %v", err)
+	}
+
+	// check that the tokens are correct
+	assert.Equal(t, encoder.EosToken, Token(6669))
+	assert.Equal(t, encoder.BosToken, Token(10989))
+	assert.Equal(t, encoder.PadToken, Token(5428))
+
+	// Clean up by removing the downloaded folder
```
Comment (sangstar) [inline gpt_bpe_test.go]: Comment doesn't match the code below it

File: resources/resolver.go [hunk 3]
```diff
@@ -752,17 +787,111 @@ func ResolveConfig(vocabId string, token string) (config *HFConfig,
...
-		hfConfig.PadTokenStr = &defaultTkn
+	hfConfig = *hfConfigPtr
+
+	return &hfConfig, resources, nil
+}
+
+// ResolveHFFromResources
+// Given a set of resources, resolve the HuggingFace configuration.
+// Used to be able to resolve both embedded and local resources.
+func ResolveHFFromResources(resources *Resources, hfConfig HFConfig) (*HFConfig, error) {
```
Comment (sangstar) [inline resources/resolver.go]: Why are you modifying `hfConfig` this way? Why not just consume `*hfConfig` for `ResolvedHFFromResources` and do something like this?

```
	hfConfig, err := ResolveHFFromResources(resources, &hfConfig)
	if err != nil {
		return nil, nil, err

	return hfConfig, resources, nil

**Review by sangstar — CHANGES_REQUESTED** [review]

File: resources/resolver.go [hunk 4]
```diff
@@ -79,6 +79,15 @@ func (rsrcs *Resources) Cleanup() {
 	}
 }
 
+// return file
+func (rsrc *Resources) GetFile(name string) (interface{}, error) {
```
Comment (sangstar) [inline resources/resolver.go:83]:
```suggestion 
func (rsrc *Resources) GetFile[T any](name string) (T, error) {
```

**Review by sangstar — CHANGES_REQUESTED** [review]

**Comment by rtalaricw** [comment]
@Rexwang8 This file is missing causing a test to fail. Do you know if it got removed?

`=== RUN   TestNerdstashEncoder_Encode2
    gpt_bpe_test.go:543: open resources/subset.jsonl: no such file or directory
    gpt_bpe_test.go:597: pass: 0, fail: 0
--- FAIL: TestNerdstashEncoder_Encode2 (0.00s)`

File: gpt_bpe_test.go (hunk 1 above)
Comment (Rexwang8) [reply gpt_bpe_test.go:863]: You're right, if it's deferred, we won't need to call it again.

File: resources/resolver.go (hunk 3 above)
Comment (Rexwang8) [reply resources/resolver.go]: You're right, it does look a little funky here, I changed it such that it follows this format and tested it, looks better

```
func ResolveHFFromResources(resources *Resources, hfConfig *HFConfig) (*HFConfig, error) {

hfConfig, err = ResolveHFFromResources(resources, hfConfig)
	if err != nil {
		return nil, nil, err
	}
```

**Comment by Rexwang8** [comment]
> @Rexwang8 This file is missing causing a test to fail. Do you know if it got removed?
> 
> `=== RUN TestNerdstashEncoder_Encode2 gpt_bpe_test.go:543: open resources/subset.jsonl: no such file or directory gpt_bpe_test.go:597: pass: 0, fail: 0 --- FAIL: TestNerdstashEncoder_Encode2 (0.00s)`

Regarding this file, I believe it was a github LFS file, It wasn't present when I started working on this batch of changes, and I don't think I modified that section.

File: resources/resolver.go (hunk 4 above)
Comment (Rexwang8) [reply resources/resolver.go:83]: interface{} is functionally equal to the go1.18 "any", considering we already use interface{} in the codebase, I'm going to opt to not change it. When we port gpt_bpe to go1.18 in the future, we can just do a quick find-and-replace to drop in any, which would be pretty clean.

**Pull request merged**